### DISCUSSION TITLE
Unable to import Gradle build

### DISCUSSION THREAD

**Discussion by niket-goel** [scalameta/metals#3007] [Q&A]
Hi,

I am new to Scala/Java development and want to use VIM (as I have with C for many years) for my dev work.
I have followed all instructions (multiple times) listed under https://scalameta.org/metals/docs/editors/vim/. 

I cannot get the coc-metals extension to work (the Coc extensions for JAVA and python work for me).

Whenever I try to follow a symbol in a scala file I get the following error:
```[coc.nvim] definition not found```

While looking through the `.metals/metals.log` I see the following error:
```
2021.07.26 14:28:40 ERROR Unexpected error when compiling clients: 'error: option -source cannot be used together with --release'
2021.07.26 16:16:08 ERROR code navigation does not work for the file '/Users/***/Workspace/code/****' because the SemanticDB file '/Users/***/Workspace/code/***/.bloop/core/build/bloop-bsp-clients-classes/classes-Metals-AlHgP7w5Q1ybap7hXnD9Sg==/META-INF/semanticdb/core/***.scala.semanticdb' doesn't exist. There can be many reasons for this error.
``` 

I have removed both `.bloop` and `.metals` directories multiple times but that did not help (expected).

I am fairly new to COC and out of ideas on where to look. 

One clue I have is that I am unable to find a `metals-vim` binary anywhere. I tried following instructions to generate the file using coursier, but I keep seeing the following errors which I am unable to resolve.
```
❯ ./coursier bootstrap --java-opt -Xss4m --java-opt -Xms100m --java-opt -Dmetals.client=coc.nvim org.scalameta:metals_2.12:0.7.7-SNAPSHOT -r bintray:scalacenter/releases -r sonatype:snapshots -o /usr/local/bin/metals-vim -f

Resolution error: Error downloading org.scalameta:metals_2.12:0.7.7-SNAPSHOT
  not found: /Users/****/.ivy2/local/org.scalameta/metals_2.12/0.7.7-SNAPSHOT/ivys/ivy.xml
  not found: https://repo1.maven.org/maven2/org/scalameta/metals_2.12/0.7.7-SNAPSHOT/metals_2.12-0.7.7-SNAPSHOT.pom
  forbidden: https://dl.bintray.com/scalacenter/releases/org/scalameta/metals_2.12/0.7.7-SNAPSHOT/metals_2.12-0.7.7-SNAPSHOT.pom
  not found: https://oss.sonatype.org/content/repositories/snapshots/org/scalameta/metals_2.12/0.7.7-SNAPSHOT/metals_2.12-0.7.7-SNAPSHOT.pom
``` 
Here is the output of CocInfo for me:
```
vim version: NVIM v0.4.4
node version: v16.3.0
coc.nvim version: 0.0.80-b427b2ccfa
coc.nvim directory: /Users/****/.local/share/nvim/plugged/coc.nvim
term: tmux
platform: darwin

```

Any help on this would be appreciated.

**Comment by ckipp01** [comment]
Hey! @niket-goel. From looking at your error message:

```
2021.07.26 14:28:40 ERROR Unexpected error when compiling clients: 'error: option -source cannot be used together with --release'
```
It looks like Metals (Bloop) isn't able to compile your code. When Bloop can't compile your code, then semanticDB can't get produced, and then navigation won't work. If your project public? If not, can you share what your javac/scalac options are? It seems that `-source` and `--release` can't be used together so I'm assuming somewhere you have these both set and you can't.

Once this is taken care of, then compilation can happen, semanticDB can be produce, and you should be able to navigate.

>. One clue I have is that I am unable to find a metals-vim binary anywhere. I tried following instructions to generate the file using coursier, but I keep seeing the following errors which I am unable to resolve

So `coc-metals` actually takes care of downloading everything for you. So there actually isn't a binary that you can locate, but it rather resolves everything and puts it together for you.  So you shouldn't have to worry at all with the bootstrap script you tried. However, I have no idea where you got that one as it's trying to pull in an _extremely_ old version of Metals.

**Reply by niket-goel** [reply]
Hey @ckipp01, thanks for your reply.

The project I am trying to run the extension on is Apache Kafka (https://github.com/apache/kafka).

I am very new to java/scala and not that familiar with their build systems yet, so I will have to dig a little to pin-point exactly where these options are set.. A cursory search through the build.gradle file for the project does yield this comment -- https://github.com/apache/kafka/blob/trunk/build.gradle#L230. 

This seems inline with the error we are seeing, right?

PS anecdotally, in the same project the coc-java extension works fine. The navigation/completion fails in and out of scala files only.

**Reply by ckipp01** [reply]
> This seems inline with the error we are seeing, right?

Correct. What version of the JVM are you using? Are you able to test with 8?

**Reply by niket-goel** [reply]
Ok.. I tried reverting my JAVA version to 1.8. 
```
2021.07.30 07:07:24 INFO  > Configure project :
2021.07.30 07:07:24 INFO  Starting build with version 7.1.0-0-ce using Gradle 7.0.2, Java 1.8 and Scala 2.13.6
```
That does fix the semantic DB error in the log, however symbol navigation is still not working for the most part.

Now I see the following warnings in the log:
(there are some other errors around bloopInstall failure, but they are more towards file already exists, so I am ignoring them. These are after freshly cleaning the .metals and .bloop dirs)
```
2021.07.30 07:08:13 WARN  no build target for: /Users/***/Workspace/code/***/core/src/main/scala/kafka/controller/KafkaController.scala
2021.07.30 07:19:05 WARN  no build target for: /Users/***/Workspace/code/***/core/src/main/scala/kafka/Kafka.scala
```

EDIT:

Just noticed this error in VIM on startup (before importing the build)
```
[coc.nvim] The "java" server crashed 5 times in the last 3 minutes. The server will not be restarted.
```

**Reply by ckipp01** [reply]
> there are some other errors around bloopInstall failure, but they are more towards file already exists, so I am ignoring them. These are after freshly cleaning the .metals and .bloop dirs

What type of errors are you seeing? If the BloopInstall fails, you'll for sure have issues. Since if it fails, it won't be able to correctly identify all the build targets in the Kafka build. This is probably why you are seeing the warnings about no build target.

> [coc.nvim] The "java" server crashed 5 times in the last 3 minutes. The server will not be restarted.

I don't believe this is coming from coc-metals. Is it possible that both coc-metals and coc-java are both being triggered here and starting up?

**Reply by niket-goel** [reply]
Here is the full log from metals. From what I can tell, the bloop installation failed because bloop was already installed.. Maybe it ran twice?

```
2021.07.30 07:28:40 ERROR
2021.07.30 07:28:40 ERROR FAILURE: Build failed with an exception.
2021.07.30 07:28:40 ERROR
2021.07.30 07:28:40 ERROR * What went wrong:
2021.07.30 07:28:40 INFO
2021.07.30 07:28:40 INFO  Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
2021.07.30 07:28:40 INFO  Use '--warning-mode all' to show the individual deprecation warnings.
2021.07.30 07:28:40 INFO  See https://docs.gradle.org/7.0.2/userguide/command_line_interface.html#sec:command_line_warnings
2021.07.30 07:28:40 INFO  10 actionable tasks: 10 executed
2021.07.30 07:28:40 ERROR Execution failed for task ':ce-http-server:bloopInstall'.
2021.07.30 07:28:40 ERROR > java.nio.file.FileAlreadyExistsException: /****/.bloop
2021.07.30 07:28:40 ERROR
2021.07.30 07:28:40 ERROR * Try:
2021.07.30 07:28:40 ERROR Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
2021.07.30 07:28:40 ERROR
2021.07.30 07:28:40 ERROR * Get more help at https://help.gradle.org
2021.07.30 07:28:40 ERROR
2021.07.30 07:28:40 ERROR BUILD FAILED in 9s
2021.07.30 07:28:40 INFO  gradle bloopInstall exit: 1
2021.07.30 07:28:40 INFO  time: ran 'gradle bloopInstall' in 9.91s
2021.07.30 07:28:40 ERROR Gradle command failed: /***/gradlew --console=plain --init-script /var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/metals5160271418348325334/init-script.gradle bloopInstall
2021.07.30 07:28:40 INFO  Attempting to connect to the build server...
Starting the bsp launcher for bloop...
Opening a bsp server connection with 'bsp --protocol local --socket /var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher3890447872190044988/bsp.socket'...
Waiting for the bsp connection to come up...
[D] Waiting for a connection at local:///var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher3890447872190044988/bsp.socket...
The server is listening for incoming connections at local:///var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher3890447872190044988/bsp.socket...
Starting thread that pumps stdin and redirects it to the bsp server...
Starting thread that pumps server stdout and redirects it to the client stdout...
2021.07.30 07:28:40 INFO  tracing is disabled for protocol BSP, to enable tracing of incoming and outgoing JSON messages create an empty file at /***/Library/Caches/org.scalameta.metals/bsp.trace.json
2021.07.30 07:28:40 INFO  time: Connected to build server in 0.19s
2021.07.30 07:28:40 INFO  Connected to Build server: Bloop v1.4.8-19-4d9f966b
2021.07.30 07:28:42 INFO  time: indexed workspace in 1.65s

```

As for the other error, we can ignore it for now if you think it's not coming out of coc-metals.. I am not sure if coc-java and coc-metals are starting up at the same time TBH.

**Reply by ckipp01** [reply]
So I transfered this over to the main Metals repo @niket-goel since I'm getting the same issue with multiple clients, not just coc-metals. I'm not fully sure what's going on as I'm getting the same errors that point towards the Bloop Install process not being able to work on the Kafka repo.

```
2021.07.31 18:47:30 INFO  Starting build with version 3.1.0-SNAPSHOT using Gradle 7.1.1, Java 1.8 and Scala 2.13.6
2021.07.31 18:47:33 INFO
2021.07.31 18:47:33 INFO  > Task :configureBloopInstall
2021.07.31 18:47:33 INFO  > Task :connect:configureBloopInstall
2021.07.31 18:47:33 INFO  > Task :connect:bloopInstall FAILED
2021.07.31 18:47:33 INFO  > Task :bloopInstall
2021.07.31 18:47:36 INFO  > Task :clients:configureBloopInstall
2021.07.31 18:47:36 INFO  > Task :examples:configureBloopInstall
2021.07.31 18:47:36 INFO  > Task :generator:configureBloopInstall
2021.07.31 18:47:36 INFO  > Task :server-common:configureBloopInstall
2021.07.31 18:47:36 INFO  > Task :raft:configureBloopInstall
2021.07.31 18:47:37 INFO  > Task :log4j-appender:configureBloopInstall
2021.07.31 18:47:37 INFO  > Task :metadata:configureBloopInstall
2021.07.31 18:47:37 INFO  > Task :jmh-benchmarks:configureBloopInstall
2021.07.31 18:47:39 INFO  > Task :shell:configureBloopInstall
2021.07.31 18:47:39 INFO  > Task :core:configureBloopInstall
2021.07.31 18:47:39 INFO
2021.07.31 18:47:39 INFO  Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
2021.07.31 18:47:39 INFO
2021.07.31 18:47:40 ERROR
2021.07.31 18:47:40 ERROR FAILURE: Build failed with an exception.
2021.07.31 18:47:40 ERROR
2021.07.31 18:47:40 ERROR * What went wrong:
2021.07.31 18:47:40 ERROR Execution failed for task ':connect:bloopInstall'.
2021.07.31 18:47:40 ERROR > java.nio.file.FileAlreadyExistsException: /Users/ckipp/Documents/scala-workspace/kafka/.bloop
2021.07.31 18:47:39 INFO  You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
2021.07.31 18:47:40 ERROR
2021.07.31 18:47:39 INFO
2021.07.31 18:47:40 ERROR * Try:
2021.07.31 18:47:40 ERROR Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
2021.07.31 18:47:40 ERROR
2021.07.31 18:47:40 ERROR * Get more help at https://help.gradle.org
2021.07.31 18:47:40 ERROR
2021.07.31 18:47:39 INFO  See https://docs.gradle.org/7.1.1/userguide/command_line_interface.html#sec:command_line_warnings
2021.07.31 18:47:40 ERROR BUILD FAILED in 17s
```
Possibly one of the others might be more familiar with Gradle and know what the issue is. However, I believe the issue is on the Bloop side of things.

**Reply by tgodzik** [reply]
Just checked newest kafka and it seems you need to first run `./gradlew compileScala` and it should generate some additional classes needed in the Scala modules.

I haven't had any issues with the import itself however. You might try to replace the script path in the Metals settings to be the local directory gradlew. Just setting `gradlew` would probably work.

**Reply by niket-goel** [reply]
Thanks for the reply @tgodzik.

I am doing a full build of the project prior to running the extensions on the code every time. I tried doing the explicit scala compile and that does not help either. 

Still getting the same errors in the `metals.log` file.

```
2021.08.02 08:36:13 INFO  See https://docs.gradle.org/7.1.1/userguide/command_line_interface.html#sec:command_line_warnings
2021.08.02 08:36:13 INFO  10 actionable tasks: 10 executed
2021.08.02 08:36:13 ERROR Execution failed for task ':connect:bloopInstall'.
2021.08.02 08:36:13 ERROR > java.nio.file.FileAlreadyExistsException: /Users/ngoel/Workspace/code/apache-kafka/kafka-fork/kafka/.bloop
2021.08.02 08:36:13 ERROR
2021.08.02 08:36:13 ERROR * Try:
2021.08.02 08:36:13 ERROR Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
2021.08.02 08:36:13 ERROR
2021.08.02 08:36:13 ERROR * Get more help at https://help.gradle.org
2021.08.02 08:36:13 ERROR
2021.08.02 08:36:13 ERROR BUILD FAILED in 6s
2021.08.02 08:36:13 INFO  gradle bloopInstall exit: 1
2021.08.02 08:36:13 INFO  time: ran 'gradle bloopInstall' in 6.86s
2021.08.02 08:36:13 ERROR Gradle command failed: /***/kafka/gradlew --console=plain --init-script /var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/metals617080777511375884/init-script.gradle bloopInstall
2021.08.02 08:36:13 INFO  Attempting to connect to the build server...
Starting the bsp launcher for bloop...
Opening a bsp server connection with 'bsp --protocol local --socket /var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher7524435036126974221/bsp.socket'...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
No server running at 127.0.0.1:8212, let's fire one...
Resolving ch.epfl.scala:bloop-frontend_2.12:1.4.8-19-4d9f966b...
Starting bloop server at 127.0.0.1:8212...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
Attempting a connection to the server...
[D] Loading 1 projects from '/***/kafka/.bloop'...
[D] Loading project from '/***/kafka/.bloop/kafka.json'
[D] Missing analysis file for project 'kafka'
[D] Waiting for a connection at local:///var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher7524435036126974221/bsp.socket...
```

>I haven't had any issues with the import itself however. You might try to replace the script path in the Metals settings to be the local directory gradlew. Just setting gradlew would probably work.

Sorry, but I am fairly new to this ecosystem. Which settings are you sugesting here? I do not think I have tweaked `metals` settings yet.

**Reply by tgodzik** [reply]
> java.nio.file.FileAlreadyExistsException: /Users/ngoel/Workspace/code/apache-kafka/kafka-fork/kafka/.bloop

This is super weird, what if you remove the `.bloop` directory ?
@Arthurm1 Did you ever see anything like this? There should not be any issues if `.bloop` already exists

> Sorry, but I am fairly new to this ecosystem. Which settings are you sugesting here? I do not think I have tweaked metals settings yet.

Not sure how to set LSP settings in Vim, but I am talking about `metals.gradleScript` setting

**Reply by Arthurm1** [reply]
@tgodzik  I don't get this issue myself with Kafka and WSL2 so I suspect it's an issue with mutli-threading

@niket-goel
Try editing Kafka's `kafka/gradle.properties` and removing the line `org.gradle.parallel=true`.
Then delete the `.bloop` dir and re-try Metals `Import Build`

With `parallel=true` the Bloop plugin will try and import all the sub-projects at once.  This greatly speeds up an import of a multi-project build but I'm not sure how well Gradle handles this when there are project inter-dependencies.  Looks like we don't handle the creation of the `.bloop` dir - which should at least be fixable.

**Comment by Arthurm1** [comment]
It looks as though we are always setting `-source` and `-target` on `bloopInstall` when we shouldn't if `--release` is present.

As a workaround you can just remove the following from the `build.gradle` file...
```groovy
    if (JavaVersion.current().isJava9Compatible())
      options.compilerArgs << "--release" << minJavaVersion
```

I'll look to fixing

**Comment by Arthurm1** [comment]
At one point you'll have to run `./gradlew processMessages processTestMessages` manually.  Bloop/Metals doesn't yet have support for source code generation.

You shouldn't need to do a full build - only this.

**Reply by niket-goel** [reply]
@Arthurm1, thanks for all your suggestions. I have tried all of it.

Changing the parallel build option and removing the release flag [1] makes the bloop install go through, which is great:
```
2021.08.03 09:31:11 INFO  BUILD SUCCESSFUL in 15s
2021.08.03 09:31:11 INFO  122 actionable tasks: 122 executed
2021.08.03 09:31:11 INFO  gradle bloopInstall exit: 0
2021.08.03 09:31:11 INFO  time: ran 'gradle bloopInstall' in 15s
2021.08.03 09:31:11 INFO  Attempting to connect to the build server...
Starting the bsp launcher for bloop...
Opening a bsp server connection with 'bsp --protocol local --socket /var/folders/vk/7_n6wxd5159bv0btnrvt07140000gq/T/bsp-launcher13319457941436170664/bsp.socket'...
Waiting for the bsp connection to come up...
Waiting for the bsp connection to come up...
. . . 
2021.08.03 09:31:12 INFO  time: Connected to build server in 1.27s
2021.08.03 09:31:12 INFO  Connected to Build server: Bloop v1.4.8-19-4d9f966b
2021.08.03 09:31:14 INFO  time: Imported build in 1.48s
2021.08.03 09:32:17 INFO  time: indexed workspace in 1m2s
```

I think Metals is working find with these changes, 


@ckipp01 -- Even with metals working fine, coc-vim is unable to provide code navigation on any scala file in the Kafka Project. Do you have any ideas on what else we can try? or is grep going to be my buddy again :) 

PS I am back to this being the first error in the metals.log:
```
2021.08.03 09:45:29 ERROR code navigation does not work for the file '/Users/***core/src/main/scala/kafka/controller/KafkaController.scala' because the SemanticDB file '/Users/***/.bloop/core/build/bloop-bsp-clients-classes/classes-Metals-mKXrZg-pQcGvm9BM8yZ8aQ==/META-INF/semanticdb/core/src/main/scala/kafka/controller/KafkaController.scala.semanticdb' doesn't exist. There can be many reasons for this error.
```

[1]
```
Comment out from build.gradle:
if (JavaVersion.current().isJava9Compatible())
      options.compilerArgs << "--release" << minJavaVersion
```

```
diff --git a/gradle.properties b/gradle.properties
--- a/gradle.properties
+++ b/gradle.properties
@@ -27,4 +27,4 @@ version=7.1.0-0-ce
 scalaVersion=2.13.6
 task=build
 org.gradle.jvmargs=-Xmx2g -Xss4m -XX:+UseParallelGC
-org.gradle.parallel=true
+org.gradle.parallel=false
```

**Reply by tgodzik** [reply]
Looks like the code hasn't compiled, do you get any diagnostics? And di you run `./gradlew processMessages processTestMessages` manuall?

**Reply by niket-goel** [reply]
>Looks like the code hasn't compiled, do you get any diagnostics? And di you run ./gradlew processMessages processTestMessages manuall?

Yes, I did run that command manually [1]. The above log is all I see when I try navigation in VIM on scala files (java navigation is working fine in the same project).

(PS i redid all steps again and still see the same error)

[1]
```
❯ ./gradlew clean processMessages processTestMessages
Starting a Gradle Daemon, 1 busy and 1 incompatible Daemons could not be reused, use --status for details

> Configure project :
Starting build with version 7.1.0-0-ce using Gradle 7.0.2, Java 11 and Scala 2.13.6
<==<===----------> 24% EXECUTING [18s]

BUILD SUCCESSFUL in 38s
71 actionable tasks: 46 executed, 25 up-to-date
```

**Reply by tgodzik** [reply]
Hmm.... at this point I was able to get the project compiled, do you see anything in the logs about compiling core or clients?

**Reply by niket-goel** [reply]
Yes.. I do see this in the `metals` log - 
```
2021.08.03 10:03:32 INFO  time: Connected to build server in 0.67s
2021.08.03 10:03:32 INFO  Connected to Build server: Bloop v1.4.8-19-4d9f966b
2021.08.03 10:03:35 INFO  time: Imported build in 1.54s
2021.08.03 10:04:40 INFO  time: indexed workspace in 1m5s
2021.08.03 10:04:40 INFO  compiling ce-resource-names (5 java sources)
2021.08.03 10:04:40 INFO  compiling ce-events-api (4 java sources)
2021.08.03 10:04:42 INFO  time: compiled ce-events-api in 1.52s
2021.08.03 10:04:42 INFO  compiling clients (1447 java sources)
2021.08.03 10:04:42 INFO  time: compiled ce-resource-names in 2.15s
2021.08.03 10:04:55 INFO  time: compiled clients in 12s

```

**Reply by tgodzik** [reply]
Looks like it's not trying to compile the core project, so maybe there is some additional source generated. Maybe you could try running the full compilation also?

There should be something in diagnostics causing the core not to be compiled.

**Reply by niket-goel** [reply]
>Looks like it's not trying to compile the core project, so maybe there is some additional source generated. Maybe you could try running the full compilation also?

Did a full compile, but getting the same error in the metals log.

> There should be something in diagnostics causing the core not to be compiled.

Which diagnostic are you referring to? I can check. (there is nothing that I could find so far TBH.

**Reply by ckipp01** [reply]
> @ckipp01 -- Even with metals working fine, coc-vim is unable to provide code navigation on any scala file in the Kafka Project. Do you have any ideas on what else we can try? or is grep going to be my buddy again :)

So again the error about missing semanticDB happens when the file either hasn't been compiled and therefore wasn't able to produce semanticDB.

> Which diagnostic are you referring to? I can check. (there is nothing that I could find so far TBH.

You should be able (if you're using the example mappings) do a `<space> a` to show all the workspace diagnostics. Do you see anything potentially causing it not to compile in there?
